unified_planning/engines/compilers/grounder.py

```diff
@@ -32,6 +32,152 @@
+    (with the :func:`~unified_planning.engines.compilers.GroundingSupport.ground_action` method).
+
+    This class also caches the grounded actions so it avoids duplication.
+    """
+
+    def __init__(
+        self,
+        problem: Problem,
+        grounding_actions_map: Optional[Dict[Action, List[Tuple[FNode, ...]]]] = None,
+    ):
+        assert isinstance(problem, Problem)
+        self._problem = problem
+        self._grounding_actions_map = grounding_actions_map
+        # grounded_actions is a map from an Action of the original problem and it's parameters
+        # to the grounded instance of the Action with the given parameters.
+        self._grounded_actions: Dict[
+            Tuple[Action, Tuple[FNode, ...]], Optional[Action]
+        ] = {}
+        env = problem.env
+        self._substituter = Substituter(env)
+        self._simplifier = Simplifier(env, problem)
+        self._grounding_result: Optional[CompilerResult] = None
+
+    @property
+    def name(self):
+        return "grounder"
+
+    def ground_action(
+        self, action: Action, parameters: Tuple[FNode, ...] = tuple()
+    ) -> Optional[Action]:
+        """
+        Grounds the given `action` with the given `parameters`.
+        An `Action` is grounded when it has no :func:`parameters <unified_planning.model.Action.parameters>`.
+
+        The returned `Action` is cached, so if the same method is called twice with the same function parameters,
+        the same object is returned, and the same object will be returned in the total problem grounding, so
+        if the resulting `Action` or :class:`~unified_planning.model.Problem` are modified, all the copies
+        returned by this class will be modified.
+
+        :param action: The `Action` that must be grounded with the given `parameters`.
+        :param parameters: The tuple of expressions used to ground the given `action`.
+        :return: The `action` grounded with the given `parameters` or `None` if the grounded
+            action does not have `effects` or the `action conditions` can be easily evaluated as a
+            contradiction.
+        """
+        assert len(action.parameters) == len(
+            parameters
+        ), "The number of given parameters for the grounding is different from the action's parameters"
+        key = (action, tuple(parameters))
+        if key in self._grounded_actions:  # The action is already created
+            return self._grounded_actions[key]
+        else:
+            subs: Dict[Expression, Expression] = dict(
+                zip(action.parameters, list(parameters))
+            )
+            new_action = create_action_with_given_subs(
+                self._problem, action, self._simplifier, self._substituter, subs
+            )
+            self._grounded_actions[key] = new_action
+            return new_action
+
+    def ground_problem(self) -> CompilerResult:
```

Comment:
This is OK, but a possibly more useful interface could be a function returning an iterator of pairs `<old_action, ground_action>`, as it can be used to easily implement a grounder but also other codes. What do you think?